### ISSUE TITLE
Texture Atlas caching for sprite sheet bundles

### ISSUE THREAD

**Issue by Trouv** [Trouv/bevy_ecs_ldtk#87]
Currently, a new texture atlas is created and added to the `Assets<TextureAtlas>` asset store as part of the logic for the `#[sprite_sheet_bundle]` field attribute of `#[derive(LdtkEntity)]`. This means that, even if two entities' visuals are part of the same tileset, two different texture atlases will be created for them, instead of them sharing the same asset with two separate asset handles.

This isn't that big of a deal, since the actual data inside a texture atlas is pretty limited. There shouldn't be any memory leakage either. But still, the current implementation isn't ideal.

Make it so that texture atlases aren't re-added to the asset store if they have been added previously. One approach could be to add an `Option<Handle<TextureAtlas>>` to the `TilesetMap`, in addition to the plain image handle. However, I am concerned that this approach may trigger `AssetEvent::Modified` when we update the field from `None` to `Some`, which could lead to unnecessary and constant reloading. An alternative approach might just search the asset store for an appropriate texture atlas handle, but this could be difficult since `TextureAtlas` isn't `Eq/PartialEq`.

**Comment by miyoku157** [comment]
I'm just looking at this issue as a beginner. But an easy solution could be to use `get_or_insert_with` instead of add?
https://docs.rs/bevy/latest/bevy/asset/struct.Assets.html#method.get_or_insert_with
Or did i understood the problem wrong?

**Comment by miyoku157** [comment]
If two different TextureAtlas can exist for the same Image, it cannot work with a hashmap, Which is what i've done currently. So i'm wondering, can the user use no_grid and grid ldtkentity together? Or is it something that should be avoid/forbid?

**Comment by Trouv** [comment]
Sorry for being slow to respond to this

> Which leads me to think that we should rethink the loading of the assets, and create the Texture atlas at the same time we loads the tileset (at least, the one configured inside ldtk json file). But it's a big change. What do you think?

Actually - I know there are some low-priority plans to make `bevy_ecs_tilemap` support using texture atlases as tilesets. This is something I've been wanting for a while, so I've considered contributing it myself but haven't gotten around to it. When this does eventually get in then it would be even more reason to just create the texture atlases as dependent assets or whatever during the asset loading process.

> If two different TextureAtlas can exist for the same Image, it cannot work with a hashmap, Which is what i've done currently. So i'm wondering, can the user use no_grid and grid ldtkentity together? Or is it something that should be avoid/forbid?

Hmm, I think `no_grid` is a case where just creating a new texture atlas on the spot is more reasonable. In the case of `no_grid` the texture atlas only has 1 rect in it, so it's not taking up a ton of space anyway.

**Comment by miyoku157** [comment]
Hmm, if that so, i think it could be a better solution to look at the bevy_ecs_tilemap support for that. Or contribute to it.
Moreover, i saw that bevy also rework `TextureAtlas` for the 0.11: https://github.com/bevyengine/bevy/pull/5103
For the `no_grid`, even if it doesn't take lots of space. If we can make multiple Texture atlas from one tileset, we can't juste handle a `Hashmap<i32, Option<Handle<TextureAtlas>>`, it has to be a set or hashset.

**Comment by Trouv** [comment]
> For the no_grid, even if it doesn't take lots of space. If we can make multiple Texture atlas from one tileset, we can't juste handle a Hashmap<i32, Option<Handle<TextureAtlas>>, it has to be a set or hashset.

Yeah that's true, I'm just saying for the `no_grid` case I don't think we need caching that badly - or we can warn users in the docs that no caching will occur.

(removing the good-first-issue label since this is actually pretty complicated)
